components/RuiExpandCollapse/src/RuiExpandCollapse.ts

```diff
@@ -0,0 +1,194 @@
+    let oldVal = this.open;
+    this._open = isOpen;
+    this.requestUpdate('open', oldVal);
+  }
+
+  /**
+   * onExpandCollapse is the handler function that is called
+   * when the user triggers an expand/collapse. This
+   * function should be overriden when trying to control
+   * the component externally
+   */
+  @property()
+  public onExpandCollapse = ():void => {
+    this.open = !this.open;
+  };
+
+  /**
+   * Internal open state of component
+   */
+  private _open: boolean = false;
+  private _collapseableEl: HTMLDivElement | null = null;
+  private _detailsSlotEl: HTMLSlotElement | null = null;
+
+  /**
+   * 
+   * The styles for the expand collapse
+   * @remarks
+   * If you are extending this class you can extend the base styles with super. Eg `return [super(), myCustomStyles]`
+   */
+  public static get styles(): CSSResultArray {
+    return [variables, layout];
+  }
+
+  /* #endregion */
+
+  /* #region Methods */
+
+  /**
+   * Handler for a click of the summary content
+   */
+  private handleClick(): void {
+    this.onExpandCollapse();
+  }
+
+  /**
+   * Sets height to 0 trigger collapse
+   * transition animation
+   */
+  private triggerCollapseAnimation(): void {
+    // add back height style and then remove on next frame to trigger animation
+    requestAnimationFrame((): void => {
+      if (this._collapseableEl) {
+        var sectionHeight = this._collapseableEl.scrollHeight;
+        this._collapseableEl.style.height = sectionHeight + 'px';
+        requestAnimationFrame((): void => {
+          if (this._collapseableEl) {
+            this._collapseableEl.style.height = 0 + 'px';
+          }
+        })
+      }
+    })
+  }
+
+  /**
+   * Sets element height to transition to,
+   * once element height is reached it unsets height
+   * style
+   */
+  private triggerExpandAnimation(): void {
+    if (this._collapseableEl) {
+      var sectionHeight = this._collapseableEl.scrollHeight;
+      this._collapseableEl.style.height = sectionHeight + 'px';
+
+      const transitionEndHandler = (): void => {
+        if (this._collapseableEl) {
+          this._collapseableEl.style.height = '';
+          this._collapseableEl.removeEventListener('transitionend', transitionEndHandler);
+        }
+      }
+  
+      this._collapseableEl.addEventListener('transitionend', transitionEndHandler);
+    }
+  }
+
+  /**
+   * Initialises the expand collapse logic and styling,
+   * once complete makes the expand collapse visible
+   */
+  private initialiseExpandCollapse(): void {
+    if (this.shadowRoot) {
+      this._collapseableEl = this.shadowRoot.querySelector('.details');
+
+      if (this._collapseableEl) {
+        
+        // need to set height initially if closed without triggering animation
+        if (!this.open) {
+          this._collapseableEl.style.height = '0px';
+        }
+        
+        let expandCollapse: HTMLElement | null = this.shadowRoot.querySelector('.expand-collapse');
```

Comment:
Can this just be inferred? 

eg `let expandCollapse = this.shadowRoot.querySelector('.expand-collapse');`

Also I'm pretty sure `querySelector` will return `undefined` not `null` if it cant find the element. Might need to test that theory.

Reply:
@jennasalau It infers the type as 'Element' so it complains on line 133 where style does not exist on type Element. 
<img width="635" alt="Screen Shot 2019-06-17 at 12 03 33 pm" src="https://user-images.githubusercontent.com/12482943/59573804-03b51900-90f8-11e9-8166-5cf7b1b3ab1b.png">
 
WRT queryselector, are you saying it always returns undefined? Or it would return undefined in this particular shadow dom case? If the former, MDN says query selector does return null on no match: https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector

Reply:
![Screen Shot 2019-06-17 at 12 15 25 pm](https://user-images.githubusercontent.com/11452925/59574179-9efabe00-90f9-11e9-97fe-328ff7b50f4f.png)

I stand corrected :D You are right.